lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -578,22 +572,256 @@ newDBLayer trace defaultFieldValues mDatabaseFile = do
+            | currentEpoch < epoch ->
+                readWalletMetadata wid
+                (delegationFromCerts Nothing Nothing Nothing)
+            | epoch + 1 < currentEpoch ->
+                readWalletMetadata wid
+                (delegationFromCerts (Just dlg) Nothing Nothing)
+            | epoch + 1 == currentEpoch ->
+                retrieveActiveDlgCert wid e dlg
+                (W.epochPred (W.EpochNo epoch))
+            | otherwise ->
+                retrieveTwoDlgsCert wid e dlg
+                (W.epochPred (W.EpochNo epoch))
+        Nothing ->
+            readWalletMetadata wid
+            (delegationFromCerts Nothing Nothing Nothing)
+
+toWalletDelegationStatus
     :: DelegationCertificate
-    -> W.DelegationDiscovered
-delegationDiscoveredFromEntity (DelegationCertificate _ slotId poolIdM) =
-    W.DelegationDiscovered slotId poolIdM
+    -> W.WalletDelegationStatus W.PoolId
+toWalletDelegationStatus (DelegationCertificate _ _ Nothing) =
+    W.NotDelegating
+toWalletDelegationStatus (DelegationCertificate _ _ (Just pool)) =
+    W.Delegating pool
 
-delegationFromEntity
+delegationFromCerts
     :: Maybe DelegationCertificate
+    -> Maybe (DelegationCertificate, W.EpochNo)
+    -> Maybe (DelegationCertificate, W.EpochNo)
     -> W.WalletDelegation W.PoolId
-delegationFromEntity = \case
-    Nothing ->
-        W.NotDelegating
-    Just (DelegationCertificate _ _ Nothing) ->
-        W.NotDelegating
-    Just (DelegationCertificate _ _ (Just pool)) ->
-        W.Delegating pool
+delegationFromCerts older interim latest = case (older,interim,latest) of
+    (Just dlg, Nothing, Nothing) ->
+        W.WalletDelegation (toWalletDelegationStatus dlg) []
+    (Just dlg1, Nothing, Just (dlg2, epoch)) ->
+        oneComingDlg dlg1 (dlg2, epoch)
+    (Just dlg1, Just (dlg2, epoch), Nothing) ->
+        oneComingDlg dlg1 (dlg2, epoch)
+    (Nothing, Nothing, Just (dlg, epoch)) ->
+        oneComingDlgAlone (dlg, epoch)
+    (Nothing, Just (dlg, epoch), Nothing) ->
+        oneComingDlgAlone (dlg, epoch)
+    (dlg, Just (dlg1, epoch1), Just (dlg2, epoch2)) ->
+        case dlg of
+            Just dlg' -> case toWalletDelegationStatus dlg' of
+                W.NotDelegating ->
+                    twoJustDlgs (dlg1, epoch1) (dlg2, epoch2)
+                W.Delegating pool ->
+                    case (toWalletDelegationStatus dlg1, toWalletDelegationStatus dlg2) of
+                        (W.NotDelegating, W.NotDelegating) ->
+                            W.WalletDelegation (W.Delegating pool) []
+                        (W.Delegating pool1, W.NotDelegating) ->
+                            comparePools pool pool1 epoch1
+                        (W.NotDelegating, W.Delegating pool1) ->
+                            comparePools pool pool1 epoch2
+                        (W.Delegating pool1, W.Delegating pool2)
+                            | pool == pool1 && pool == pool2 ->
```

Comment:
so here, I just assumed that if we have situation that we delegating to poolId, and then in one epoch X there we a number of certificates with the latest joining to poolId, and in the X+1 epoch the same, then we do not want anything in next, as we are delegating to poolID and it is not going to be changed. This is what I feel should be presented (if we do not outlaw this kind of stuff which was already subject of another comment). Below are similar situations taken into consideration